File: mytravelstatus/mobile/ios/MyTravelStatus/Services/PrivacySettings.swift
/**
 * PrivacySettings.swift
 *
 * Manages user privacy preferences and settings.
 * Ensures all optional features are opt-in with clear user consent.
 *
 * @package MyTravelStatus
 * @since   1.0.0
 */

import Foundation
import Combine

@MainActor
class PrivacySettings: ObservableObject {
    static let shared = PrivacySettings()

    // MARK: - Keys

    private enum Keys {
        static let backgroundLocationEnabled = "privacy.backgroundLocationEnabled"
        static let backgroundLocationEducationShown = "privacy.backgroundLocationEducationShown"
        static let analyticsEnabled = "privacy.analyticsEnabled"
        static let analyticsEducationShown = "privacy.analyticsEducationShown"
        static let crashReportingEnabled = "privacy.crashReportingEnabled"
        static let photoMetadataConfirmed = "privacy.photoMetadataConfirmed"
        static let calendarAccessConfirmed = "privacy.calendarAccessConfirmed"
        static let lastPrivacyPolicyAccepted = "privacy.lastPolicyAccepted"
        static let demoModeEnabled = "privacy.demoModeEnabled"
    }

    // MARK: - Published Properties

    /// Background location tracking - OFF by default, opt-in only
    @Published var backgroundLocationEnabled: Bool {
        didSet {
            UserDefaults.standard.set(backgroundLocationEnabled, forKey: Keys.backgroundLocationEnabled)
            if !backgroundLocationEnabled {
                // Disable background updates when user opts out
                BackgroundLocationManager.shared.isTrackingEnabled = false
            }
        }
    }

    /// Whether we've shown the background location education screen
    @Published var backgroundLocationEducationShown: Bool {
        didSet {
            UserDefaults.standard.set(backgroundLocationEducationShown, forKey: Keys.backgroundLocationEducationShown)
        }
    }

    /// Analytics collection - OFF by default, opt-in only
    @Published var analyticsEnabled: Bool {
        didSet {
            UserDefaults.standard.set(analyticsEnabled, forKey: Keys.analyticsEnabled)
            AnalyticsManager.shared.updateEnabled(analyticsEnabled)
        }
    }

    /// Whether we've shown the analytics education screen
    @Published var analyticsEducationShown: Bool {
        didSet {
            UserDefaults.standard.set(analyticsEducationShown, forKey: Keys.analyticsEducationShown)
        }
    }

    /// Crash reporting - ON by default (opt-out)
    @Published var crashReportingEnabled: Bool {
        didSet {
            UserDefaults.standard.set(crashReportingEnabled, forKey: Keys.crashReportingEnabled)
        }
    }

    /// User confirmed EXIF metadata extraction from photos
    @Published var photoMetadataConfirmed: Bool {
        didSet {
            UserDefaults.standard.set(photoMetadataConfirmed, forKey: Keys.photoMetadataConfirmed)
        }
    }

    /// User confirmed calendar scanning
    @Published var calendarAccessConfirmed: Bool {
        didSet {
            UserDefaults.standard.set(calendarAccessConfirmed, forKey: Keys.calendarAccessConfirmed)
        }
    }

    /// Demo mode for App Store reviewers
    @Published var demoModeEnabled: Bool {
        didSet {
            UserDefaults.standard.set(demoModeEnabled, forKey: Keys.demoModeEnabled)
        }
    }

    /// Date of last privacy policy acceptance
    @Published var lastPrivacyPolicyAccepted: Date? {
        didSet {
            if let date = lastPrivacyPolicyAccepted {
                UserDefaults.standard.set(date.timeIntervalSince1970, forKey: Keys.lastPrivacyPolicyAccepted)
            }
        }
    }

    // MARK: - Initialization

    private init() {
        // Load saved preferences - defaults ensure privacy-first approach
        self.backgroundLocationEnabled = UserDefaults.standard.bool(forKey: Keys.backgroundLocationEnabled)
        self.backgroundLocationEducationShown = UserDefaults.standard.bool(forKey: Keys.backgroundLocationEducationShown)

        // Analytics OFF by default
        self.analyticsEnabled = UserDefaults.standard.object(forKey: Keys.analyticsEnabled) as? Bool ?? false
        self.analyticsEducationShown = UserDefaults.standard.bool(forKey: Keys.analyticsEducationShown)

        // Crash reporting ON by default (non-tracking)
        self.crashReportingEnabled = UserDefaults.standard.object(forKey: Keys.crashReportingEnabled) as? Bool ?? true

        self.photoMetadataConfirmed = UserDefaults.standard.bool(forKey: Keys.photoMetadataConfirmed)
        self.calendarAccessConfirmed = UserDefaults.standard.bool(forKey: Keys.calendarAccessConfirmed)
        self.demoModeEnabled = UserDefaults.standard.bool(forKey: Keys.demoModeEnabled)

        if let timestamp = UserDefaults.standard.object(forKey: Keys.lastPrivacyPolicyAccepted) as? TimeInterval {
            self.lastPrivacyPolicyAccepted = Date(timeIntervalSince1970: timestamp)
        } else {
            self.lastPrivacyPolicyAccepted = nil
        }
    }

    // MARK: - Public Methods

    /// Reset all privacy settings to defaults
    func resetToDefaults() {
        backgroundLocationEnabled = false
        analyticsEnabled = false
        crashReportingEnabled = true
        photoMetadataConfirmed = false
        calendarAccessConfirmed = false
        // Don't reset education flags - user has seen them
    }

    /// Clear all data for account deletion
    func clearAllData() {
        UserDefaults.standard.removeObject(forKey: Keys.backgroundLocationEnabled)
        UserDefaults.standard.removeObject(forKey: Keys.backgroundLocationEducationShown)
        UserDefaults.standard.removeObject(forKey: Keys.analyticsEnabled)
        UserDefaults.standard.removeObject(forKey: Keys.analyticsEducationShown)
        UserDefaults.standard.removeObject(forKey: Keys.crashReportingEnabled)
        UserDefaults.standard.removeObject(forKey: Keys.photoMetadataConfirmed)
        UserDefaults.standard.removeObject(forKey: Keys.calendarAccessConfirmed)
        UserDefaults.standard.removeObject(forKey: Keys.lastPrivacyPolicyAccepted)
        UserDefaults.standard.removeObject(forKey: Keys.demoModeEnabled)

        // Reset in-memory values
        backgroundLocationEnabled = false
        backgroundLocationEducationShown = false
        analyticsEnabled = false
        analyticsEducationShown = false
        crashReportingEnabled = true
        photoMetadataConfirmed = false
        calendarAccessConfirmed = false
        demoModeEnabled = false
        lastPrivacyPolicyAccepted = nil
    }

    /// Accept privacy policy
    func acceptPrivacyPolicy() {
        lastPrivacyPolicyAccepted = Date()
    }

    /// Check if privacy policy needs re-acceptance (e.g., after update)
    var needsPrivacyPolicyAcceptance: Bool {
        guard let accepted = lastPrivacyPolicyAccepted else { return true }
        // Policy was updated on this date - require re-acceptance
        let policyUpdateDate = ISO8601DateFormatter().date(from: "2026-01-01T00:00:00Z") ?? Date.distantPast
        return accepted < policyUpdateDate
    }

    // MARK: - Data Summary for Export

    /// Generate summary of data collected for privacy report
    func getDataCollectionSummary() -> [String: Any] {
        return [
            "settings": [
                "backgroundLocationEnabled": backgroundLocationEnabled,
                "analyticsEnabled": analyticsEnabled,
                "crashReportingEnabled": crashReportingEnabled,
                "photoMetadataConfirmed": photoMetadataConfirmed,
                "calendarAccessConfirmed": calendarAccessConfirmed
            ],
            "dataCollected": [
                "location": backgroundLocationEnabled ? "Country-level only (3x daily)" : "Manual entries only",
                "photos": photoMetadataConfirmed ? "GPS metadata from selected photos" : "Not accessed",
                "calendar": calendarAccessConfirmed ? "Event titles and locations" : "Not accessed",
                "analytics": analyticsEnabled ? "Anonymous usage events" : "Disabled",
                "crashes": crashReportingEnabled ? "Anonymous crash reports" : "Disabled"
            ],
            "dataSharedWithThirdParties": false,
            "dataUsedForTracking": false
        ]
    }
}

File: mytravelstatus/mobile/android/app/src/main/java/com/mytravelstatus/app/service/PrivacySettings.kt
/**
 * PrivacySettings.kt
 *
 * Manages user privacy preferences and consent.
 * - Background location: OFF by default, opt-in only
 * - Analytics: OFF by default, opt-in only
 * - Crash reporting: ON by default, opt-out
 *
 * Aligned with Google Play Data Safety requirements.
 *
 * @package MyTravelStatus
 * @since   1.0.0
 */

package com.mytravelstatus.app.service

import android.content.Context
import android.content.SharedPreferences
import androidx.core.content.edit
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

object PrivacySettings {
    private const val PREFS_NAME = "privacy_settings"

    // Preference keys
    private object Keys {
        const val BACKGROUND_LOCATION_ENABLED = "background_location_enabled"
        const val BACKGROUND_LOCATION_EDUCATION_SHOWN = "background_location_education_shown"
        const val ANALYTICS_ENABLED = "analytics_enabled"
        const val CRASH_REPORTING_ENABLED = "crash_reporting_enabled"
        const val PHOTO_EXIF_CONFIRMED = "photo_exif_confirmed"
        const val NOTIFICATION_PERMISSION_REQUESTED = "notification_permission_requested"
        const val DEMO_MODE_ENABLED = "demo_mode_enabled"
        const val PRIVACY_POLICY_VERSION_ACCEPTED = "privacy_policy_version_accepted"
    }

    private var prefs: SharedPreferences? = null

    // StateFlows for reactive UI updates
    private val _backgroundLocationEnabled = MutableStateFlow(false)
    val backgroundLocationEnabled: StateFlow<Boolean> = _backgroundLocationEnabled.asStateFlow()

    private val _analyticsEnabled = MutableStateFlow(false)
    val analyticsEnabled: StateFlow<Boolean> = _analyticsEnabled.asStateFlow()

    private val _crashReportingEnabled = MutableStateFlow(true)
    val crashReportingEnabled: StateFlow<Boolean> = _crashReportingEnabled.asStateFlow()

    private val _demoModeEnabled = MutableStateFlow(false)
    val demoModeEnabled: StateFlow<Boolean> = _demoModeEnabled.asStateFlow()

    // Current privacy policy version
    const val CURRENT_PRIVACY_POLICY_VERSION = "1.0.0"

    /**
     * Initialize privacy settings
     */
    fun init(context: Context) {
        if (prefs != null) return

        prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

        // Load initial values from preferences
        // IMPORTANT: Background location defaults to FALSE (opt-in)
        _backgroundLocationEnabled.value = prefs?.getBoolean(Keys.BACKGROUND_LOCATION_ENABLED, false) ?: false
        // IMPORTANT: Analytics defaults to FALSE (opt-in)
        _analyticsEnabled.value = prefs?.getBoolean(Keys.ANALYTICS_ENABLED, false) ?: false
        // Crash reporting defaults to TRUE (opt-out)
        _crashReportingEnabled.value = prefs?.getBoolean(Keys.CRASH_REPORTING_ENABLED, true) ?: true
        // Demo mode defaults to FALSE
        _demoModeEnabled.value = prefs?.getBoolean(Keys.DEMO_MODE_ENABLED, false) ?: false
    }

    // MARK: - Background Location

    /**
     * Check if background location education has been shown
     */
    fun hasShownBackgroundLocationEducation(): Boolean {
        return prefs?.getBoolean(Keys.BACKGROUND_LOCATION_EDUCATION_SHOWN, false) ?: false
    }

    /**
     * Mark background location education as shown
     */
    fun markBackgroundLocationEducationShown() {
        prefs?.edit { putBoolean(Keys.BACKGROUND_LOCATION_EDUCATION_SHOWN, true) }
    }

    /**
     * Set background location enabled status
     * NOTE: This should only be called AFTER showing the education screen
     */
    fun setBackgroundLocationEnabled(enabled: Boolean) {
        prefs?.edit { putBoolean(Keys.BACKGROUND_LOCATION_ENABLED, enabled) }
        _backgroundLocationEnabled.value = enabled
    }

    // MARK: - Analytics

    /**
     * Set analytics enabled status
     */
    fun setAnalyticsEnabled(enabled: Boolean) {
        prefs?.edit { putBoolean(Keys.ANALYTICS_ENABLED, enabled) }
        _analyticsEnabled.value = enabled
    }

    // MARK: - Crash Reporting

    /**
     * Set crash reporting enabled status
     */
    fun setCrashReportingEnabled(enabled: Boolean) {
        prefs?.edit { putBoolean(Keys.CRASH_REPORTING_ENABLED, enabled) }
        _crashReportingEnabled.value = enabled
    }

    // MARK: - Photo EXIF

    /**
     * Check if user has confirmed photo EXIF extraction
     */
    fun hasConfirmedPhotoExif(): Boolean {
        return prefs?.getBoolean(Keys.PHOTO_EXIF_CONFIRMED, false) ?: false
    }

    /**
     * Mark photo EXIF extraction as confirmed
     */
    fun markPhotoExifConfirmed() {
        prefs?.edit { putBoolean(Keys.PHOTO_EXIF_CONFIRMED, true) }
    }

    /**
     * Reset photo EXIF confirmation (for next session)
     */
    fun resetPhotoExifConfirmation() {
        prefs?.edit { putBoolean(Keys.PHOTO_EXIF_CONFIRMED, false) }
    }

    // MARK: - Notifications

    /**
     * Check if notification permission has been requested
     */
    fun hasRequestedNotificationPermission(): Boolean {
        return prefs?.getBoolean(Keys.NOTIFICATION_PERMISSION_REQUESTED, false) ?: false
    }

    /**
     * Mark notification permission as requested
     */
    fun markNotificationPermissionRequested() {
        prefs?.edit { putBoolean(Keys.NOTIFICATION_PERMISSION_REQUESTED, true) }
    }

    // MARK: - Demo Mode

    /**
     * Set demo mode enabled status
     */
    fun setDemoModeEnabled(enabled: Boolean) {
        prefs?.edit { putBoolean(Keys.DEMO_MODE_ENABLED, enabled) }
        _demoModeEnabled.value = enabled
    }

    // MARK: - Privacy Policy

    /**
     * Get the accepted privacy policy version
     */
    fun getAcceptedPrivacyPolicyVersion(): String? {
        return prefs?.getString(Keys.PRIVACY_POLICY_VERSION_ACCEPTED, null)
    }

    /**
     * Accept the current privacy policy version
     */
    fun acceptPrivacyPolicy() {
        prefs?.edit { putString(Keys.PRIVACY_POLICY_VERSION_ACCEPTED, CURRENT_PRIVACY_POLICY_VERSION) }
    }

    /**
     * Check if user needs to accept updated privacy policy
     */
    fun needsPrivacyPolicyAcceptance(): Boolean {
        val accepted = getAcceptedPrivacyPolicyVersion()
        return accepted != CURRENT_PRIVACY_POLICY_VERSION
    }

    // MARK: - Data Collection Summary

    /**
     * Get a summary of current data collection settings for export
     */
    fun getDataCollectionSummary(): Map<String, Any> {
        return mapOf(
            "backgroundLocationEnabled" to _backgroundLocationEnabled.value,
            "analyticsEnabled" to _analyticsEnabled.value,
            "crashReportingEnabled" to _crashReportingEnabled.value,
            "demoModeEnabled" to _demoModeEnabled.value,
            "privacyPolicyVersionAccepted" to (getAcceptedPrivacyPolicyVersion() ?: "none")
        )
    }

    // MARK: - Clear All Data

    /**
     * Clear all privacy settings (for account deletion)
     */
    fun clearAllData() {
        prefs?.edit { clear() }

        // Reset StateFlows to defaults
        _backgroundLocationEnabled.value = false
        _analyticsEnabled.value = false
        _crashReportingEnabled.value = true
        _demoModeEnabled.value = false
    }
}
